newton/_src/sim/builder.py

```
@@ -4968,125 +4979,460 @@ def add_rod(
+        edge_v: list[int] = []
+        edge_len: list[float] = []
+        edge_bodies: list[int] = []
+
+        # Create all edge bodies first.
+        for e_idx, (u, v) in enumerate(edges):
+            if u < 0 or u >= num_nodes or v < 0 or v >= num_nodes:
                 raise ValueError(
-                    f"add_rod: segment {i} has a too-small length (length={float(segment_length):.3e}); "
-                    f"segment length must be > {min_segment_length:.1e}"
+                    f"add_rod_graph: edge {e_idx} has invalid node indices ({u}, {v}) for {num_nodes} nodes"
                 )
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)